manual/manual/tutorials/polymorphism.etex

```diff
@@ -0,0 +1,443 @@
+defined as
+\begin{caml_example}
+  type 'a nested = List of 'a list | Nested of 'a list nested;;
+\end{caml_example}
+Intuitively, a value of type "'a nested" is a list of list \dots of list of
+elements "a" with "k" nested list. An interesting function on this type would
+be the "depth" function that compute this "k". As a first try, we can define
+\begin{caml_example}
+let rec depth = function
+  | List _ -> 1
+  | Nested n -> 1 + depth n;;
+\end{caml_example}
+The type error here comes from the fact that during the definition of "depth",
+the type checker first assign to it the type "'a -> 'b ".
+When typing the pattern matching, it becomes "'a nested -> 'b" then
+"'a nested -> int" once the "List" branch is typed.
+However, when typing the application "depth n" in the "Nested" branch,
+the type checker encounter a problem: "depth n" is applied to
+"'a list nested", it must therefore have the type
+"'a list nested -> 'b", unifying this constraint with the previous one,
+leads to the impossible constraint "'a list nested = 'a nested"
+
+In other words, within its definition, the recursive function "depth" is
+applied to values of type "'a t" with different types "'a". This creates a
+problem because the type checker introduces new type variable "'a" only at the
+\emph{definition} of the function "depth" whereas, here, it should use a
+different type variable for every \emph{application} of the function "depth".
+
+\subsection{Explicitly polymorphic annotations}
+The solution of this conundrum is to use an explicitly polymorphic type
+annotation for the type "'a":
+\begin{caml_example}
+let rec depth: 'a. 'a nested -> int = function
+  | List _ -> 1
+  | Nested n -> 1 + depth n;;
+depth ( Nested(List [ [7]; [8] ]) );;
+\end{caml_example}
+In the type of "depth",  "'a.'a nested -> int", the type variable "'a"
+is universally quantified, i.e "'a.'a nested -> int" reads as
+``for all type "'a", "depth" maps "'a nested" values to integers''.
+Whereas the standard type "'a nested -> int" can be interpreted
+as ``let be a type variable "'a", then "depth" maps "'a nested" values
+to integers''. There is two major differences with these two type
+expression. First, the explicit polymorphic annotation indicates to the
+type checker that it needs to introduce a new type variable every times
+the function depth is applied. This solves our problem with the definition
+of the function depth.
+
+Second, it also notifies the type checker that the type of the function should
+be polymorphic. Indeed, without explicit polymorphic type annotation, the
+following type annotation is perfectly valid
+\begin{caml_example}
+  let sum: 'a -> 'b -> 'c = fun x y -> x + y;;
+\end{caml_example}
+since "'a" denotes a type variable that may or may not be polymorphic.
+Whereas, it is an error to unify an explicitly polymorphic type with a
+non-polymorphic type:
+\begin{caml_example}
+  let sum: 'a 'b 'c. 'a -> 'b -> 'c = fun x y -> x + y;;
+\end{caml_example}
```

Comment:
> First, the practical comment that many types in explicitly annotated functions can be simply left as "_", e.g. let rec depth : 'a . 'a nested -> _, limiting the annotation effort.

Good point.

> Second, point out that there is also a variation using "newtypes" (let rec depth : type t . t nested -> _) which often yields the same effect though the treatment of the variable inside the function is slightly different when GADTs are involved (i.e. the variable can be specialized differently depending on the GADT case whereas the normal type variable 'a means always the same type on the current recursion level).

It is certainly an important point to make and this is the right place to do so, but I fear this not the right time since GADT and locally abstract type are still relegated to the extension section of the manual.